zookeeper/client.go

```diff
@@ -14,26 +14,45 @@ import (
 type StateListener func(state ClientState)
 
 type Client struct {
-	conn        *zk.Conn
+	conn        ZKConnection
 	acl         []zk.ACL
 	basePath    string
 	nodeOwner   string
 	nodeSchema  string
 	zkState     zk.State
 	clientState ClientState
-	listeners   []StateListener
+	listeners   map[string]StateListener
 	sync.Mutex
 }
 
 type ZKClient interface {
 	Close()
 	ClientState() ClientState
-	RegisterListener(listener StateListener)
+	RegisterListener(id string, listener StateListener)
+	UnregisterListener(id string)
 
-	Exists(path string) (bool, error)
-	Get(path string) ([]byte, error)
+	Exists(path string) (bool, int32, error)
+	existsW(path string) (bool, int32, <-chan zk.Event, error)
+	Get(path string) ([]byte, int32, error)
+	getW(path string) ([]byte, int32, <-chan zk.Event, error)
 	Create(path string, data []byte, perms []int32) error
-	Set(path string, data []byte) error
+	Set(path string, data []byte) (int32, error)
+	Children(path string) ([]string, int32, error)
+	childrenW(path string) ([]string, int32, <-chan zk.Event, error)
+}
+
+type ZKConnection interface {
+	AddAuth(scheme string, auth []byte) error
+	Children(path string) ([]string, *zk.Stat, error)
+	ChildrenW(path string) ([]string, *zk.Stat, <-chan zk.Event, error)
```

Comment:
Is `*W` stands for `Watch`? Somekind of convention? :) I know that in `Go` even the inventor is fond of one letter variables, should we follow them tho?

Reply:
`*W` does stand for watch. This particular interface just matches the connection struct from `github.com/samuel/go-zookeeper`, but we could name our private functions on the client `*Watch` instead. My idea for making those private on the client is that they are only seen/used by the `ValueNodeWatcher` & `ParentNodeWatcher`.

I also initially had those calling the `zk.Conn` directly but decided it better to refactor and always interface with the connection through the client.

Reply:
ok, let's keep it 👍 thanks!